packages/deployments/contracts/contracts/messaging/libraries/Queue.sol

```diff
@@ -104,21 +107,60 @@ library QueueLib {
 
     bytes32[] memory items = new bytes32[](last + 1 - first);
     uint256 index; // Cursor for index in the batch of `items`.
+    uint256 removedCount; // If any items have been removed, we filter them here.
     // NOTE: `first <= last` rephrased here to `!(first > last)` as it's a cheaper condition.
     while (!(first > last)) {
-      items[index] = queue.data[first];
+      bytes32 item = queue.data[first];
+      // Check to see if the item has been removed before appending it to the array.
+      if (!queue.removed[item]) {
+        items[index] = item;
+        unchecked {
+          ++index;
+        }
+      } else {
+        // The item was removed. We do NOT increment the index (we will re-use this position).
+        unchecked {
+          ++removedCount;
+        }
+      }
+
       // Delete the item and the commitBlock.
+      // NOTE: We do NOT delete the entry from `queue.removed`, as it's a reverse lookup and we want to
+      // block that value permanently (e.g. if there's multiple of the same bad value in the queue).
       delete queue.data[first];
       delete queue.commitBlock[first];
 
       unchecked {
-        ++index;
         ++first;
       }
     }
+
     // Update the value for `first` in our queue object since we've dequeued a number of elements.
     queue.first = first;
-    return items;
+
+    if (removedCount == 0) {
+      return items;
+    } else {
+      // If some items were removed, there will be a number of trailing 0 values we need to truncate
+      // from the array. Create a new array with all of the items up until these empty values.
+      uint256 amendedLength = last + 1 - first - removedCount;
```

Comment:
This will fail since first has now incremented to last+1 due to earlier while loop.
I think we could use amendedLength=index since index contains the actual number of items being inserted